MSC3758: Add `event_property_is` push rule condition kind (#3758)
matrix-org/matrix-spec-proposals: proposals/3758-expand-push-rule-conditions.md

> # MSCXXXX: Expand push rule conditions

Reviewer: ```suggestion
# MSC3758: Expand push rule conditions
```

Also, please wrap your lines at ~80-120 characters

Author: https://github.com/matrix-org/matrix-spec-proposals/pull/3758/commits/a10d5a2b91a4d7d2e5eaaf8cf9b4745ef30d3778

# MSC3758: Expand push rule conditions

Currently the only condition used to match event content for push rules is the `event_match` kind.
This compares a glob-style string against a string value within the event dictionary. The event
dictionary is flattened before conditions are checked to enable mathing nested values.

This approach is currently limited to only checking for string values within the event dictionary
(at any level). This MSC proposes a new exact match type for event content that works with all
JSON types.

## Proposal

### Exact matching event data

We propose a new type of condition, `exact_event_match`. Similar to  the current `event_match`
([link to spec](https://spec.matrix.org/v1.3/client-server-api/#conditions-1)), this condition

> takes two parameters: `value` and `key`. The exact match compares the `value` to the event data

Reviewer 2: It could potentially be useful for a `content` rule to opt-into this [since `content` rules are essentially limited `event_match` rules](https://spec.matrix.org/v1.4/client-server-api/#push-rules):

> Content rules take one parameter: `pattern`, that gives the glob pattern to match against. This is treated in the same way as `pattern` for `event_match`.

Reviewer 2: This could be a future extension. 🤷

associated with `key` exactly. Both type and content (when a string) should be identical. This allows

> for matching all valid JSON types `null`, `true`, `false` and numbers. This also provides a simpler

Reviewer 2: Earlier it says:

> This MSC proposes a new exact match type for event content that works with all JSON types.

And then here we list out:

> for matching all valid JSON types `null`, `true`, `false` and numbers.

Is the intention of this only to match "simple" values or would it be valid to do something like:

```json
{
  "kind": "exact_event_match",
  "key": "event.content.is_something",
  "value": []
}
```

Reviewer 2: Also, I think this is limited to integers since non-integers are (nominally) forbidden? (Depending on your room version and such, but I don't think we should build in features around that.)

Author: Ah originally lists were included but later split into [MSC3887](https://github.com/matrix-org/matrix-spec-proposals/pull/3887) so this is targets just those simple types. Are non-integers forbidden? I know we have some custom bridge events that contain booleans 😅, but wasn't aware of spec that limited content fields!

Reviewer 2: I somehow completely missed that MSC, which is essentially the other thing I need for MSC3952...

Reviewer 3: *Unresolving*

Can we get MSC3887 mentioned in the MSC text please? Also, neither this MSC or MSC3887 appear to handle JSON objects - this should be called out as not supported, or supported.

Reviewer 2: ```suggestion
associated with `key` exactly. Both type and content (when a string) should be identical
(include case). This allows for matching all non-compound JSON types allowed by
[canonical JSON](https://spec.matrix.org/v1.5/appendices/#canonical-json):
i.e. strings, `null`, `true`, `false` and integers. This also provides a simpler
```

exact string matching mechanism (and any associated performance gains on implementation side without
globbing).

An example condition may look like (encoded as a JSON object):

```json
{
  "kind": "exact_event_match",
  "key": "event.content.is_something",
  "value": true
}
```

## Alternatives

[MSC3862](https://github.com/matrix-org/matrix-spec-proposals/pull/3862) proposes an alternative
solution by converting non-string JSON objects to strings in the `event_match` condition type.

> ## Security considerations
> 
> None.

Reviewer 2: ```suggestion
## Security considerations

None.

## Future extensions

A future MSC may wish to define the behavior of `exact_event_match` when
used with a JSON object or array.

[MSC3887](https://github.com/matrix-org/matrix-spec-proposals/pull/3887) is a
related MSC which attempts to define behavior for searching for a value inside of
an array.
```

> # MSC3758: Expand push rule conditions

Reviewer 3: This MSC appears to be missing a few sections from the template which would answer some questions about what else was considered, etc :)

> # MSC3758: Expand push rule conditions

Reviewer 3: (We'll need signoff on this MSC as well, ftr)

Reviewer 2: This has now been done, see description!

Author: Done!

Currently the only condition used to match event content for push rules is the `event_match` kind.
This compares a glob-style string against a string value within the event dictionary. The event
dictionary is flattened before conditions are checked to enable mathing nested values.

This approach is currently limited to only checking for string values within the event dictionary
(at any level). This MSC brings makes two proposals to enable matching event content of any other
JSON type within push rules. [MSC1796](https://github.com/matrix-org/matrix-spec-proposals/pull/1796)
is an example change that would require this extended matching.

## Proposals

### Exact matching event data

> We propose a new type of condition, `exact_event_match`. Similar to `event_match`, this condition

Reviewer 3: I'd suggest we go as far to deprecate `event_match`, tbh.

Author: I think there's some use in being able to glob match event content, even if not used in the baserules I believe it's possible to do via custom rules?

Reviewer 4: You can glob match using globs already. You don't need the spec to add implicit globs on either end, if you didn't provide the globs yourself.

Reviewer 2: I think deprecating that should be done as a separate step, if it is wanted, to avoid further bikeshedding on this MSC.

takes two parameters: `pattern` and `key`. The exact match compares the `pattern` to the event data

> associated with `key` exactly. Both type and content (when a string) should be identical. This allows

Reviewer 5: Given that an _exact_ match is done, I think the term "pattern" is a bit misleading. Maybe something like "value" could be clearer?

Author: Agreed, updated.

for matching all valid JSON types `null`, `true`, `false` and numbers. This also provides a simpler
exact string matching mechanism (and any associated performance gains on implementation side without
globbing).

An example condition may look like (encoded as a JSON object):

```json
{
  "kind": "exact_event_match",

>   "key": "event.content.is_something",

Reviewer 6: I'm bound to say that I'd be reluctant for us to specify a new condition without us also fixing https://github.com/matrix-org/matrix-spec/issues/648.

Reviewer 6: (though maybe having both `event_match` *and* `exact_event_match` with *different* semantics for `key` is even worse?

Bah every time I look at push rules I end up just wanting to burn the whole system to the ground)

Reviewer 5: We can possibly solve this separately through https://github.com/matrix-org/matrix-spec-proposals/pull/3873 and then reapply this or whatever other solution is chosen for `exact_event_match`.

Author: I think keeping the behaviour the same as `event_match` is less confusing than changing it, although I totally agree this is very annoying! #3873 seems like a reasonable workaround for this.

  "pattern": true
}
```

> ### Handling of lists within event data

Author: Agreed, I have removed the list content from this MSC and started [MSC3887](https://github.com/matrix-org/matrix-spec-proposals/pull/3887).

We propose that conditons simply match any value within a list. Conditions should be evaluated against
every item in the list and return a match if any of the items meet the condition. For example an event:

```json
{
  "content": {
    "mentions": ["@user-a:beeper.com", "@user-b:beeper.com"]
  }
}
```

The following condition would match:

```json
{
  "kind": "event_match",
  "key": "event.content.mentions",
  "pattern": "@user-a:beeper.com"
}

> ```

Reviewer 6: >  Specifically, I can see servers serializing lists down to a string and running the regex over that string, which could mean excessive matches for quotes, commas, and other list serialization commonalities.

.... well, don't do that then?

> ```

Reviewer 3: correct, but also we should consider what the average server developer is probably going to do with a given spec.

> ```

Reviewer 2: > Specifically, I can see servers serializing lists down to a string and running the regex over that string, which could mean excessive matches for quotes, commas, and other list serialization commonalities.

Empirically while implementing it that never crossed my mind. The implementation at matrix-org/synapse#14964 doesn't handle lists anyway, only strings, ints, booleans, and null.

> # MSC3758: Expand push rule conditions

Reviewer: ```suggestion
# MSC3758: Add `exact_event_match` push rule condition kind 
```

for consistency with the MSC title, and to be more descriptive

> # MSC3758: Add `exact_event_match` push rule condition kind 

Reviewer 2: I would have to ask @Fizzadar, but I assume this was based on [the `event_match` condition](https://spec.matrix.org/v1.5/client-server-api/#conditions-1), but it being exact, not globby. Maybe that was misnamed from the start and we shouldn't propagate that though!

> # MSC3758: Add `exact_event_match` push rule condition kind 

Author: Ha! This is totally based on `event_match` and I'd support it being renamed `exact_field_match` 👍

> # MSC3758: Add `exact_event_match` push rule condition kind 

Reviewer 2: I suspect none of us really care what it is called since it is just a constant -- I'd also be fine with `exact_field_match`. (Although I think I was recently told to use "property" instead of "field", so maybe `exact_property_match`?)

> # MSC3758: Add `exact_event_match` push rule condition kind 

Reviewer 2: I settled on `event_property_is`, hopefully that's clearer is? 🤷

This compares a glob-style string against a string value within the event dictionary. The event
dictionary is flattened before conditions are checked to enable matching nested values.

This approach is currently limited to only checking for string values within the event dictionary
(at any level). This MSC proposes a new exact match type for event content that works with all
JSON types.

## Proposal

### Exact matching event data

We propose a new type of condition, `exact_event_match`. Similar to  the current `event_match`
([link to spec](https://spec.matrix.org/v1.3/client-server-api/#conditions-1)), this condition
takes two parameters: `value` and `key`. The exact match compares the `value` to the event data
associated with `key` exactly. Both type and content (when a string) should be identical
(include case). This allows for matching all non-compound JSON types allowed by
[canonical JSON](https://spec.matrix.org/v1.5/appendices/#canonical-json):
i.e. strings, `null`, `true`, `false` and integers. This also provides a simpler
exact string matching mechanism (and any associated performance gains on implementation side without

> globbing).

Reviewer 7: What's the intention for event_match? Do they both continue to live on for different purposes? Seems like this is just a replacement though, in which case shouldn't this MSC deprecate event_match?

Reviewer 2: It is not a replacement -- `event_match` does globbing (and case insensitive matching). This is useful for keyword matching, if nothing else.

Reviewer 2: @dbkr Do you want me to make this clearer in the MSC? I feel like it is pretty clear from the preamble, but I can more explicitly call out it does not deprecate `event_match`.

Reviewer 7: It does seem fairly clear actually. probably OK.
